[feat] add analytics for payments

mumer92 opened the pull request:
### Description

[LEARNER-8823](https://openedx.atlassian.net/browse/LEARNER-8823)

Analytics for IAP

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -33,9 +35,17 @@ class CourseUpgradeHelper: NSObject {
+        case error(PurchaseError, Error?)
+    }
+    
+    enum ErrorAction: String {
+        case refreshToRetry = "refresh to retry"
+        case emailSupport = "email support"
+        case close = "close"
```
Can you specify what data is only used for the analytics? this will improve the readability of the code.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +55,55 @@ class CourseUpgradeHelper: NSObject {
+    private var paymentVerifyTime: CFTimeInterval?
+    
+    private var environment: Environment?
+    private var pacing: String?
+    private var courseID: CourseBlockID?
+    private var blockID: CourseBlockID?
+    private var screen: CourseUpgradeScreen = .none
+    private var coursePrice: String?
+    
+    private var isRefresh = false
```
The data is being saved in a shared object, it's a good idea to refresh/delete the data on the success/error so we don't have invalid data copy at any moment.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +55,55 @@ class CourseUpgradeHelper: NSObject {
-        
+    
+    private var startTime: CFTimeInterval?
+    private var paymentVerifyTime: CFTimeInterval?
```
I guess it's also a start time for verifying the payment.

mumer92 replied:
`paymentVerifyTime` measures when OS dialog is started and measures time when a response is returned.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +55,55 @@ class CourseUpgradeHelper: NSObject {
-        case .error(let type):
+        case .error(let type, let error):
+            if type == .paymentError, let error = error {
+                environment?.analytics.trackCourseUpgradePaymentError(courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", coursePrice: coursePrice ?? "", screen: screen, paymentError: error.localizedDescription)
+            } else {
+                environment?.analytics.trackCourseUpgradeError(courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", coursePrice: coursePrice ?? "", screen: screen, paymentError: type.rawValue)
```
The `paymentError` value is different in `Upgrade Error` and `Upgrade Error`. Shouldn't it be the same?

mumer92 replied:
i suppose one is the error that is returned by OS payment dialog and other is errors defined in response to our api

saeedbashir replied:
We also have a local copy for that, `PaymentError`.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -80,24 +128,33 @@ class CourseUpgradeHelper: NSObject {
+            alertController.addButton(withTitle: Strings.CourseUpgrade.FailureAlert.refreshToRetry, style: .default) { [weak self] _ in
+                self?.trackUpgradeErrorAction(errorAction: ErrorAction.refreshToRetry.rawValue)
+                self?.isRefresh = true
+                self?.startTime = CFAbsoluteTimeGetCurrent()
```
The `startTime` is being updated here, shouldn't this will affect the value for `Time to Unlock Upgraded Content Intial Try`.

saeedbashir commented on Source/CourseUnknownBlockViewController.swift:
```diff
@@ -203,28 +208,34 @@ class CourseUnknownBlockViewController: UIViewController, CourseBlockViewControl
                 self?.enableUserInteraction(enable: true)
                 upgradeView.updateUpgradeButtonVisibility(visible: false)
-                self?.courseUpgradeHelper.handleCourseUpgrade(state: .success(self?.courseID ?? "", self?.blockID), screen: .courseUnit)
+                self?.courseUpgradeHelper.handleCourseUpgrade(state: .complete)
+                self?.courseUpgradeHelper.handleCourseUpgrade(state: .success(courseID, self?.blockID))
```
Can we merge both of these statements?

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +57,67 @@ class CourseUpgradeHelper: NSObject {
+        self.screen = screen
+    }
+    
+    func clearData() {
```
From where you are calling it?

mumer92 replied:
added callee.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +57,67 @@ class CourseUpgradeHelper: NSObject {
+            startTime = CFAbsoluteTimeGetCurrent()
+            break
+        case .payment:
+            paymentVerifyTime = CFAbsoluteTimeGetCurrent()
```
Shouldn't this time be started from the `fulfillment` because that's the state when the execute api is hit?

mumer92 replied:
this is for the payment dialog that appears.

saeedbashir replied:
But the name `Payments: Time to Verify Payment` says otherwise. It seems you are tying to capture how much time is bing taken for verifying(execute api) the payment.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +57,67 @@ class CourseUpgradeHelper: NSObject {
+        case .payment:
+            paymentVerifyTime = CFAbsoluteTimeGetCurrent()
+            break
         case .fulfillment:
```
Typo: `fulfilment`.

mumer92 replied:
`fulfillment` is correct.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +57,67 @@ class CourseUpgradeHelper: NSObject {
             showLoader()
             break
+        case .complete:
+            environment?.analytics.trackUpgradeNow(with: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", screenName: screen)
```
Any reason to track `trackUpgradeNow` on completion of the process?

mumer92 replied:
added by mistake, removed.

saeedbashir commented on Source/OEXAnalytics+Swift.swift:
```diff
@@ -83,6 +83,15 @@ public enum AnalyticsDisplayName : String {
     case ProfileDeleteAccountClicked = "Profile: Delete Account Clicked"
     case RegistrationOptinTurnedOn = "Registration: Opt-in Turned On"
     case RegistrationOptinTurnedOff = "Registration: Opt-in Turned Off"
+    case CourseUpgradeNowSuccess = "Upgrade Now Success"
+    case CourseUpgradeSuccessDuration = "Time to Unlock Upgraded Content Intial Try"
+    case CourseUpgradeSuccessDurationAfterRefresh = "Time to Unlock Upgraded Content After Refresh"
+    case CourseUpgradeTimeToVerifyPayment = "Time to Verify Payment"
+    case CourseUpgradeTimeToLoadPrice = "Time to Load Price"
+    case CourseUpgradePaymentError = "Payment Error"
+    case CourseUpgradeError = "Upgrade Error"
+    case CourseUpgradeLoadError = "Load Error"
+    case CourseUpgradeErrorAction = "Error Action"
```
I guess the event names have been updated.

saeedbashir commented on Source/OEXAnalytics+Swift.swift:
```diff
@@ -620,5 +641,137 @@ extension OEXAnalytics {
+        let event = OEXAnalyticsEvent()
+        event.displayName = AnalyticsDisplayName.CourseUpgradeNowSuccess.rawValue
+        event.name = AnalyticsEventName.CourseUpgradeNowSuccess.rawValue
+        
```
Can we ask the android team and set the event category as well?

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,6 +23,8 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
     
+    typealias Environment = OEXAnalyticsProvider
+    
     static let courseID = "CourseID"
     static let blockID = "BlockID"
     static let screen = "Screen"
```
These are unsed and can be removed.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -33,9 +35,18 @@ class CourseUpgradeHelper: NSObject {
+        case close = "close"
     }
 
     private var upgradeModel: CourseUpgradeModel?
```
Can use only one from `upgradeModel` & `courseUpgradeModel` with private setter and public getter.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +56,60 @@ class CourseUpgradeHelper: NSObject {
+        self.screen = screen
+    }
+    
+    func clearData() {
```
Can you check is it safe to also call it from `resetUpgradeModel` method?

saeedbashir reviewed (changes requested):
* `environment.analytics.trackUpgradeNow(with: courseID, blockID: courseSku, pacing: pacing, screenName: screen)` from `ValuePropDetailViewController.swift` you are passing `courseSku` as blockID is it intentional? 
* App is crashing when trying to log analytics for API Errors. I tried it by not using JWT authentication.
* I'm not sure but can you please ask Poornima or Android Team either we want to group `Upgrade Now Clicked` with `Payments` or not?
* Analytics `Value Prop Banner Clicked` is in the sheet but it's not implemented yet. Is this is intentional? 
* You are not tracking `Payments: Course Upgrade Success`.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -45,20 +56,60 @@ class CourseUpgradeHelper: NSObject {
             break
-        case .error(let type):
+        case .error(let type, let error):
+            environment?.analytics.trackCourseUpgradePaymentError(courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", coursePrice: coursePrice ?? "", screen: screen, paymentError: error?.localizedDescription ?? "")
```
I'm not sure but it seems passing the `CourseUpgradeHandler.shared.formattedError` is more accurate for error. In this way, we will be passing more details.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -72,6 +123,19 @@ class CourseUpgradeHelper: NSObject {
+            environment?.analytics.trackCourseUpgradeDuration(isRefresh: true, courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", coursePrice: coursePrice ?? "", screen: screen, elapsedTime: refreshEndTime.millisecond)
+        }
+        
+        refreshTime = nil
+        startTime = nil
```
Can it be reset in the `clearData`/resetData method?

saeedbashir commented on Source/OEXAnalytics+Swift.swift:
```diff
@@ -83,6 +84,15 @@ public enum AnalyticsDisplayName : String {
     case ProfileDeleteAccountClicked = "Profile: Delete Account Clicked"
     case RegistrationOptinTurnedOn = "Registration: Opt-in Turned On"
     case RegistrationOptinTurnedOff = "Registration: Opt-in Turned Off"
+    case CourseUpgradeNowSuccess = "Payments: Course Upgrade Success"
```
Can be renamed to `CourseUpgradeSuccess`

saeedbashir commented on Source/ValuePropDetailViewController.swift:
```diff
@@ -84,19 +90,37 @@ class ValuePropDetailViewController: UIViewController, InterfaceOrientationOverr
             PaymentManager.shared.productPrice(courseSku) { [weak self] price in
                 if let price = price {
+                    let endTime = CFAbsoluteTimeGetCurrent() - startTime
+                    self?.trackPriceLoadDuration(elapsedTime: endTime.millisecond)
+                    self?.coursePrice = price
```
You are trying to capture the analytics before setting the price. Can it be done in the setter of the `coursePrice`?

mumer92 replied:
actually the intent here is time for it takes to load the price from store, after getting price, it set to screen and time is also captured here, if we were to observe this inside the setter, have to make changes to time variable to be available out to class scope.

saeedbashir commented on Source/ValuePropDetailViewController.swift:
```diff
@@ -167,33 +191,36 @@ class ValuePropDetailViewController: UIViewController, InterfaceOrientationOverr
         
-        let pacing = course.isSelfPaced ? "self" : "instructor"
-        environment.analytics.trackUpgradeNow(with: course.course_id ?? "", blockID: courseSku, pacing: pacing)
+        environment.analytics.trackUpgradeNow(with: courseID, blockID: courseSku, pacing: pacing, screenName: screen)
```
Is it intentional of passing `courseSku` to blockID ?

mumer92 commented on Source/CourseUnknownBlockViewController.swift:
```diff
@@ -203,28 +208,33 @@ class CourseUnknownBlockViewController: UIViewController, CourseBlockViewControl
         
-        let pacing = course.isSelfPaced ? "self" : "instructor"
-        environment.analytics.trackUpgradeNow(with: course.course_id ?? "", blockID: self.blockID ?? "", pacing: pacing)
+        environment.analytics.trackUpgradeNow(with: courseID, blockID: self.blockID ?? "", pacing: pacing, screenName: .courseUnit)
```
removed from the component view class

mumer92 commented:
> `environment.analytics.trackUpgradeNow(with: courseID, blockID: courseSku, pacing: pacing, screenName: screen)` from `ValuePropDetailViewController.swift` you are passing `courseSku` as blockID is it intentional?

It was being passed previously, i have updated it now.

> * App is crashing when trying to log analytics for API Errors. I tried it by not using JWT authentication.

I was not able to reproduce this, can you share log of at which event app is crashing, that would be helpful.

> * I'm not sure but can you please ask Poornima or Android Team either we want to group `Upgrade Now Clicked` with `Payments` or not?

Added question for them.

> * Analytics `Value Prop Banner Clicked` is in the sheet but it's not implemented yet. Is this is intentional?

`Value Prop Learn More Clicked` is en event which already exists that co-insides with that, after discussion, added screenName to its param

> * You are not tracking `Payments: Course Upgrade Success`.

Added

> * We normally name BI value according to the analytics name like `Payments: Time to Unlock Upgraded Content`. Please rename the others accordingly.

Had discussion on the analytics draft sheet and changed bi names accordingly.

saeedbashir commented on Source/CourseOutlineViewController.swift:
```diff
@@ -334,7 +334,7 @@ public class CourseOutlineViewController :
     private func handleNavigationIfNeeded() {
         if let courseUpgradeModel = courseUpgradeHelper.courseUpgradeModel {
-            courseUpgradeHelper.resetUpgradeModel()
+            courseUpgradeHelper.clearData()
```
I guess it will impact the analytics that you will send from removal of the view like `Payments: Time to Unlock Upgraded Content` because the data won't be there to send in the analytics.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,55 +23,106 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
     private override init() { }
     
-    func handleCourseUpgrade(state: CompletionState, screen: CourseUpgradeScreen, delegate: CourseUpgradeHelperDelegate? = nil) {
+    func setupCourse(environment: Environment, pacing: String, courseID: CourseBlockID, blockID: CourseBlockID? = nil, coursePrice: String, screen: CourseUpgradeScreen) {
```
It's not setting up the course but instead setting up the required data. name it accordingly.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,55 +23,106 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
+    // These times are being in analytics
+    private var startTime: CFTimeInterval?
+    private var refreshTime: CFTimeInterval?
+    private var timeForiOSPaymentModal: CFTimeInterval?
```
It will be for the SDK payment time. Name it accordingly.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,55 +23,106 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
             break
-        case .error(let type):
+        case .error(let type, _):
+            environment?.analytics.trackCourseUpgradePaymentError(courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", coursePrice: coursePrice ?? "", screen: screen, paymentError: CourseUpgradeHandler.shared.formattedError)
```
I guess tracking of two analytics required here, `Payments: Payment Error` and `Payments: Course Upgrade Error` depending on the error but you are not tracking `Payments: Course Upgrade Error` any where.

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,55 +23,106 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
+            environment?.analytics.trackCourseUpgradeDuration(isRefresh: true, courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", coursePrice: coursePrice ?? "", screen: screen, elapsedTime: refreshEndTime.millisecond)
+        }
+        
+        environment?.analytics.trackCourseUpgradeSuccess(courseID: courseID ?? "", blockID: blockID ?? "", pacing: pacing ?? "", price: coursePrice ?? "", screen: screen)
```
How about calling the `clearData()` at the end of this success method?

saeedbashir commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -80,33 +131,40 @@ class CourseUpgradeHelper: NSObject {
                 self?.unlockController.removeView() {
                     self?.delegate?.hideAlertAction()
-                    self?.resetUpgradeModel()
+                    self?.clearData()
```
Although it's fine  but how about moving it along with `hideAlertAction` into a method and calling that instead because it's being called from all the places. Thoughts?

saeedbashir commented on Source/OEXAnalytics+Swift.swift:
```diff
@@ -156,6 +166,15 @@ public enum AnalyticsEventName: String {
     case ProfileDeleteAccountClicked = "edx.bi.app.profile.delete_account.clicked"
     case RegistrationOptinTurnedOn = "edx.bi.app.user.register.opt_in.on"
     case RegistrationOptinTurnedOff = "edx.bi.app.user.register.opt_in.off"
+    case CourseUpgradeSuccess = "edx.bi.payments.upgrade_now.success"
+    case CourseUpgradeSuccessDuration = "edx.bi.payments.upgrade_duration"
+    case CourseUpgradeSuccessDurationAfterRefresh = "edx.bi.payments.time_to_unlock_content_after_refresh"
+    case CourseUpgradeTimeToVerifyPayment = "edx.bi.payments.time_to_verify_payment"
+    case CourseUpgradeTimeToLoadPrice = "edx.bi.payments.time_to_load"
+    case CourseUpgradePaymentError = "edx.bi.payments.payment_error"
+    case CourseUpgradeError = "edx.bi.payments.upgrade_error"
+    case CourseUpgradeLoadError = "edx.bi.payments.load_error"
+    case CourseUpgradeErrorAction = "edx.bi.payments.error_action"
```
Some BI values are not as per the sheet like `edx.bi.payments.course_upgrade.success`, `edx.bi.payments.time_to_load_price`, `edx.bi.payments.course_upgrade_error`, `edx.bi.payments.price_load_error`, `edx.bi.payments.error_alert_action`

saeedbashir reviewed (changes requested):
Please update the event names `Payments: Upgrade Now Clicked` and `Payments: Time to Verify Payment` according the decision.

mumer92 commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,55 +23,122 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
+    
+    private(set) var courseUpgradeModel: CourseUpgradeModel?
+    
+    // These times are being in analytics
```
```suggestion
    // These times are being used in analytics
```

mumer92 commented on Source/CourseUpgradeHelper.swift:
```diff
@@ -23,55 +23,122 @@ protocol CourseUpgradeHelperDelegate: AnyObject {
+    enum ErrorAction: String {
+        case refreshToRetry = "refresh"
+        case reloadPrice = "reload_price"
+        case emailSupport = "get help"
```
how about changing it to `get_help` as well?
